MoltenVK/MoltenVK/Utility/MVKSmallVector.h

```diff
@@ -59,9 +59,10 @@ using MVKSmallVector = std::vector<T>;
 template<typename Type, typename Allocator = mvk_smallvector_allocator<Type, 0>>
 class MVKSmallVectorImpl
 {
-  Allocator  alc;
-  
+
 public:
+  Allocator  alc;
+

```

Comment:
I wonder if you shouldn't add a constructor to `MVKSmallVector<T>` and its base to initialize the allocator. Then you wouldn't have to reach into the allocator in order to set the command buffer for the bump allocator.

Reply:
Maybe, to me that seems to add complication for no clear reason, thus I'm leaving it for others to decide. Are you thinking doing something of the sort:
`_bindings = MVKCommandVector(MVKCommandVectorAllocator<BlahBlah>(cmdBuff));`
in `setContent` functions, or passing the command buffer to all MVKCommand constructors and putting
`_bindings(MVKCommandVector(MVKCommandVectorAllocator<BlahBlah>(cmdBuff)))`
in the inializer list, or something else?

Reply:
@js6i , I highly discourage this change to MVKSmallVector. The allocator should not be exposed at all and the overall design wasn't ment to be used with a specific constructor as well. The allocator type is a pure template argument.
I think the problem is, that MVKCommandVectorAllocator is already bad by design, it should not have the member cmdBuffer and it should not call mvkPushCommandMemory directly.
There are multiple ways of redesigning this, but since I am not really familiar with MoltenVK, I can't suggest a good solution right now, since I am afraid I will brake other things.
But please do not change MVKSmallVector with the above change, it will introduce all kind of issues in the future.

Reply:
@aerofly, I don't see what's the argument you're making, other than saying it's bad. Setting the allocator with accompanying data for an expandable array is a very simple thing to do, and I don't understand how that's going to be a source of issues.

Reply:
Sorry, I didn't mean to be rude here. What I tried to say is, that exposing the allocator and actually accessing allocator elements from another class is basically mixing separate concepts. The allocator as well as the actual vector should not be intertwined like this.

I have added MVKSmallVector to MoltenVK to have a std::vector compatible container that allows the easy use of pre allocated stack elements to be used, to avoid new/delete's.
The way it is used now would not be STL compatible.

Now for this special case, I recommend to write a special class that encapsulates everything that is intended here.

Reply:
Regarding the first point, I still don't get it. If you want to use some allocator for an array or whatever, you gotta set it and its things. I don't see how that's mixing concepts. If I put the pointer to command buffer directly into the vector class, then I'd agree it would be mixing unrelated things.

I don't know if we need STL compatibility for anything. If we do, that's probably a valid argument.

That said, I also don't mind having dedicated code for this.